File: backend-py/app/services/settings_csv.py
"""
Импорт / экспорт настроек пользователя в CSV (явное чтение и запись файла на диске / в поток).

Подходит под учебное требование «работа с внешними CSV» при том, что основное хранилище — MongoDB.
"""
from __future__ import annotations

import csv
import io
from typing import Any

CSV_HEADERS = [
    "preferredLanguage",
    "preferredTheme",
    "telegramDailyNotify",
    "telegramActivityNotify",
    "telegramDailyNotifyHour",
    "telegramQuietHoursEnabled",
    "telegramQuietStartHour",
    "telegramQuietEndHour",
]


def _bool_to_cell(v: Any) -> str:
    return "true" if v else "false"


def user_doc_to_row(user: dict[str, Any]) -> dict[str, str]:
    """Строит одну строку CSV из документа пользователя в MongoDB."""
    hour = user.get("telegramDailyNotifyHour")
    q_start = user.get("telegramQuietStartHour")
    q_end = user.get("telegramQuietEndHour")
    try:
        hour_n = int(hour) if hour is not None and hour != "" else 9
    except (TypeError, ValueError):
        hour_n = 9
    try:
        qs = int(q_start) if q_start is not None and q_start != "" else 22
    except (TypeError, ValueError):
        qs = 22
    try:
        qe = int(q_end) if q_end is not None and q_end != "" else 8
    except (TypeError, ValueError):
        qe = 8
    hour_n = max(0, min(23, hour_n))
    qs = max(0, min(23, qs))
    qe = max(0, min(23, qe))

    return {
        "preferredLanguage": str(user.get("preferredLanguage") or "ru"),
        "preferredTheme": str(user.get("preferredTheme") or "light"),
        "telegramDailyNotify": _bool_to_cell(user.get("telegramDailyNotify")),
        "telegramActivityNotify": _bool_to_cell(user.get("telegramActivityNotify")),
        "telegramDailyNotifyHour": str(hour_n),
        "telegramQuietHoursEnabled": _bool_to_cell(user.get("telegramQuietHoursEnabled")),
        "telegramQuietStartHour": str(qs),
        "telegramQuietEndHour": str(qe),
    }


def row_to_csv_bytes(row: dict[str, str]) -> bytes:
    """Сериализация одной строки настроек в UTF-8 CSV (для скачивания)."""
    buf = io.StringIO()
    writer = csv.DictWriter(buf, fieldnames=CSV_HEADERS, lineterminator="\n")
    writer.writeheader()
    writer.writerow({k: row.get(k, "") for k in CSV_HEADERS})
    return buf.getvalue().encode("utf-8")


def _parse_bool_cell(raw: str) -> bool:
    x = raw.strip().lower()
    if x in ("true", "1", "yes", "on"):
        return True
    if x in ("false", "0", "no", "off"):
        return False
    raise ValueError(f"Invalid boolean value in CSV: {raw!r} (use true/false)")


def _parse_hour_cell(raw: str) -> int:
    n = int(raw.strip())
    if n < 0 or n > 23:
        raise ValueError(f"Hour must be 0–23, got {n}")
    return n


def parse_csv_to_updates(text: str) -> dict[str, Any]:
    """
    Разбор CSV: первая строка данных после заголовка.
    Возвращает поля для $set в MongoDB. Пустые ячейки пропускаются.
    """
    reader = csv.DictReader(io.StringIO(text))
    if reader.fieldnames is None:
        raise ValueError("CSV has no header row")

    header_set = {h.strip() for h in reader.fieldnames if h and h.strip()}
    unknown = header_set - set(CSV_HEADERS)
    if unknown:
        raise ValueError(f"Unknown columns: {', '.join(sorted(unknown))}")

    rows = list(reader)
    if not rows:
        raise ValueError("CSV has no data rows")

    raw = rows[0]
    updates: dict[str, Any] = {}

    def get_cell(key: str) -> str | None:
        if key not in raw:
            return None
        v = raw[key]
        if v is None:
            return None
        s = str(v).strip()
        return s if s else None

    lang = get_cell("preferredLanguage")
    if lang is not None:
        lang = lang.lower()
        if lang not in ("ru", "en"):
            raise ValueError("preferredLanguage must be ru or en")
        updates["preferredLanguage"] = lang

    theme = get_cell("preferredTheme")
    if theme is not None:
        theme = theme.lower()
        if theme not in ("light", "dark"):
            raise ValueError("preferredTheme must be light or dark")
        updates["preferredTheme"] = theme

    cell = get_cell("telegramDailyNotify")
    if cell is not None:
        updates["telegramDailyNotify"] = _parse_bool_cell(cell)

    cell = get_cell("telegramActivityNotify")
    if cell is not None:
        updates["telegramActivityNotify"] = _parse_bool_cell(cell)

    cell = get_cell("telegramDailyNotifyHour")
    if cell is not None:
        updates["telegramDailyNotifyHour"] = _parse_hour_cell(cell)

    cell = get_cell("telegramQuietHoursEnabled")
    if cell is not None:
        updates["telegramQuietHoursEnabled"] = _parse_bool_cell(cell)

    cell = get_cell("telegramQuietStartHour")
    if cell is not None:
        updates["telegramQuietStartHour"] = _parse_hour_cell(cell)

    cell = get_cell("telegramQuietEndHour")
    if cell is not None:
        updates["telegramQuietEndHour"] = _parse_hour_cell(cell)

    return updates

File: backend/src/utils/settingsCsv.js
// @ts-nocheck
/** CSV export/import for user settings (mirrors backend-py/app/services/settings_csv.py). */

const CSV_HEADERS = [
  'preferredLanguage',
  'preferredTheme',
  'telegramDailyNotify',
  'telegramActivityNotify',
  'telegramDailyNotifyHour',
  'telegramQuietHoursEnabled',
  'telegramQuietStartHour',
  'telegramQuietEndHour',
];

function boolToCell(v) {
  return v ? 'true' : 'false';
}

function userDocToRow(user) {
  const hour = Number(user.telegramDailyNotifyHour);
  const qs = Number(user.telegramQuietStartHour);
  const qe = Number(user.telegramQuietEndHour);
  const hourN = Number.isFinite(hour) ? Math.max(0, Math.min(23, hour)) : 9;
  const qStart = Number.isFinite(qs) ? Math.max(0, Math.min(23, qs)) : 22;
  const qEnd = Number.isFinite(qe) ? Math.max(0, Math.min(23, qe)) : 8;

  return {
    preferredLanguage: String(user.preferredLanguage || 'ru'),
    preferredTheme: String(user.preferredTheme || 'light'),
    telegramDailyNotify: boolToCell(Boolean(user.telegramDailyNotify)),
    telegramActivityNotify: boolToCell(user.telegramActivityNotify !== false),
    telegramDailyNotifyHour: String(hourN),
    telegramQuietHoursEnabled: boolToCell(Boolean(user.telegramQuietHoursEnabled)),
    telegramQuietStartHour: String(qStart),
    telegramQuietEndHour: String(qEnd),
  };
}

function escapeCsvCell(value) {
  const s = String(value ?? '');
  if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function rowToCsvBytes(row) {
  const header = CSV_HEADERS.join(',');
  const line = CSV_HEADERS.map((k) => escapeCsvCell(row[k] ?? '')).join(',');
  return Buffer.from(`${header}\n${line}\n`, 'utf8');
}

function parseBoolCell(raw) {
  const x = String(raw).trim().toLowerCase();
  if (['true', '1', 'yes', 'on'].includes(x)) return true;
  if (['false', '0', 'no', 'off'].includes(x)) return false;
  throw new Error(`Invalid boolean value in CSV: ${raw} (use true/false)`);
}

function parseHourCell(raw) {
  const n = parseInt(String(raw).trim(), 10);
  if (!Number.isFinite(n) || n < 0 || n > 23) throw new Error(`Hour must be 0–23, got ${raw}`);
  return n;
}

function parseCsvLine(line) {
  const out = [];
  let cur = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (inQuotes) {
      if (ch === '"') {
        if (line[i + 1] === '"') {
          cur += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        cur += ch;
      }
    } else if (ch === '"') {
      inQuotes = true;
    } else if (ch === ',') {
      out.push(cur);
      cur = '';
    } else {
      cur += ch;
    }
  }
  out.push(cur);
  return out;
}

function parseCsvToUpdates(text) {
  const normalized = String(text || '').replace(/^\uFEFF/, '').trim();
  const lines = normalized.split(/\r?\n/).filter((l) => l.trim());
  if (lines.length < 2) throw new Error('CSV has no data rows');

  const headers = parseCsvLine(lines[0]).map((h) => h.trim());
  const headerSet = new Set(headers.filter(Boolean));
  const unknown = [...headerSet].filter((h) => !CSV_HEADERS.includes(h));
  if (unknown.length) throw new Error(`Unknown columns: ${unknown.sort().join(', ')}`);

  const values = parseCsvLine(lines[1]);
  const raw = {};
  headers.forEach((h, i) => {
    raw[h] = values[i] ?? '';
  });

  const updates = {};
  const getCell = (key) => {
    if (!(key in raw)) return null;
    const s = String(raw[key] ?? '').trim();
    return s || null;
  };

  const lang = getCell('preferredLanguage');
  if (lang !== null) {
    const l = lang.toLowerCase();
    if (!['ru', 'en'].includes(l)) throw new Error('preferredLanguage must be ru or en');
    updates.preferredLanguage = l;
  }

  const theme = getCell('preferredTheme');
  if (theme !== null) {
    const th = theme.toLowerCase();
    if (!['light', 'dark'].includes(th)) throw new Error('preferredTheme must be light or dark');
    updates.preferredTheme = th;
  }

  let cell = getCell('telegramDailyNotify');
  if (cell !== null) updates.telegramDailyNotify = parseBoolCell(cell);

  cell = getCell('telegramActivityNotify');
  if (cell !== null) updates.telegramActivityNotify = parseBoolCell(cell);

  cell = getCell('telegramDailyNotifyHour');
  if (cell !== null) updates.telegramDailyNotifyHour = parseHourCell(cell);

  cell = getCell('telegramQuietHoursEnabled');
  if (cell !== null) updates.telegramQuietHoursEnabled = parseBoolCell(cell);

  cell = getCell('telegramQuietStartHour');
  if (cell !== null) updates.telegramQuietStartHour = parseHourCell(cell);

  cell = getCell('telegramQuietEndHour');
  if (cell !== null) updates.telegramQuietEndHour = parseHourCell(cell);

  if (!Object.keys(updates).length) {
    throw new Error('No settings to import (fill at least one column)');
  }

  return updates;
}

module.exports = {
  CSV_HEADERS,
  userDocToRow,
  rowToCsvBytes,
  parseCsvToUpdates,
};
